### PR TITLE
scripts: Fix twisterlib for ruff

### PR DESCRIPTION
**Pull request by LukaszMrugala**
Twisterlib files are often changed and having them under ruff's linting would be useful.
This brings them up to ruff's standards and updates relevant tests.

### CODE REVIEW

**Comment by pdgendt** [comment]
I appreciate the effort! Just wondering if pushing it as a single commit is a good idea?

**Comment by LukaszMrugala** [comment]
> I appreciate the effort! Just wondering if pushing it as a single commit is a good idea?

Whoops. Didn't really check the diff size while in-progress. When doing the first few files, it seemed unnecessary, as in practice most commits would be for a single file. But now, with the diff size ballooning to c. 3k LOC, I think separating them would be prudent, even if just to make the review easier.

I mostly pushed it here to verify blackboxes in CI and will start the separation after fixing them, if necessary.

**Comment by golowanow** [comment]
I'd suggest to split test changes from code changes to reduce risk of regressions when both are changes simultaneously,
IMO better to ruff Twister tests in a separate PR first.

**Comment by nashif** [comment]
> I appreciate the effort! Just wondering if pushing it as a single commit is a good idea?

@LukaszMrugala  i think @pdgendt  meant a commit per "violation", not a commit per file :-)

**Comment by LukaszMrugala** [comment]
> I'd suggest to split test changes from code changes to reduce risk of regressions when both are changed simultaneously, IMO better to ruff Twister tests in a separate PR first.

I do not agree with doing the tests first. Currently, before we change tests, they serve as an indicator when changing the main Twister code. If we change tests first, we introduce uncertainty into our twisterlib changes for no reason.

Test changes should be coupled with the twisterlib changes that necessitate them, so no commit introduces broken code (code that fails tests).

**Comment by golowanow** [comment]
> I do not agree with doing the tests first. Currently, before we change tests, they serve as an indicator when changing the main Twister code. If we change tests first, we introduce uncertainty into our twisterlib changes for no reason.

This ruff change in Twister is a sort of refactoring without functionality changes, right ?
When you change both Twister tests and Twister code in the same (this) PR, there is risk of false positive (a bug in both simultaneously), so it is less secure.

**Review by pdgendt — COMMENTED** [review]
Nice work!

**Review by pdgendt — DISMISSED** [review]

File: scripts/pylib/twister/twisterlib/handlers.py [hunk 1]
```diff
@@ -123,8 +123,10 @@ def _verify_ztest_suite_name(self, harness_status, detected_suite_names, handler
             return
         # compare the expect and detect from end one by one without order
         _d_suite = detected_suite_names[-len(expected_suite_names):]
-        if set(_d_suite) != set(expected_suite_names):
-            if not set(_d_suite).issubset(set(expected_suite_names)):
+        if (
+            set(_d_suite) != set(expected_suite_names)
+            and not set(_d_suite).issubset(set(expected_suite_names))
+        ):
                 self._missing_suite_name(expected_suite_names, handler_time)
```
[the remark this answers is not in the document]
Comment (LukaszMrugala) [reply scripts/pylib/twister/twisterlib/handlers.py]: Good catch, not sure why linter didn't get it.
It also seemed to miss some %-based formatting, but I've noticed that in time.

**Comment by LukaszMrugala** [comment]
> This ruff change in Twister is a sort of refactoring without functionality changes, right ? When you change both Twister tests and Twister code in the same (this) PR, there is risk of false positive (a bug in both simultaneously), so it is less secure.

It should be just refactoring, but some changes are not NOPs. Take for example B006 - mutable argument default. To fix it, you need to change the code. Not to mention the biggest culprit - using context managers for file handling, which completely dissolved two whole functions.

Additionally, unit tests often directly mock functions used by the function under test. If I change that function, how would you have it so both the test passes and is not changed?

**Review by nashif — DISMISSED** [review]

**Comment by pdgendt** [comment]
Needs a rebase to resolve conflict

**Review by pdgendt — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by glenn-andrews** [comment]
This breaks testing on my on my Windows installation:

```
west twister --device-testing --device-serial COM3 --device-serial-baud 115200 -p disco_l475_iot1  -T .
Traceback (most recent call last):
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\commands.py", line 546, in __call__
    mod = _commands_module_from_file(self.py_file)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\commands.py", line 701, in _commands_module_from_file
    spec.loader.exec_module(mod)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "C:\Users\glenn\zephyrproject\zephyr\scripts\west_commands\twister_cmd.py", line 21, in <module>
    from twisterlib.twister_main import main
  File "C:\Users\glenn\zephyrproject\zephyr\scripts\pylib\twister\twisterlib\twister_main.py", line 20, in <module>
    from twisterlib.runner import TwisterRunner
  File "C:\Users\glenn\zephyrproject\zephyr\scripts\pylib\twister\twisterlib\runner.py", line 26, in <module>
    from domains import Domains
ModuleNotFoundError: No module named 'domains'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 505, in run_command
    self.run_extension(args.command, argv)
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 623, in run_extension
    self.cmd = self.extensions[name].factory()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\commands.py", line 548, in __call__
    raise ExtensionCommandError(
west.commands.ExtensionCommandError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\glenn\zephyrproject\.venv\Scripts\west.exe\__main__.py", line 7, in <module>
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 1085, in main
    app.run(argv or sys.argv[1:])
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 244, in run
    self.run_command(argv, early_args)
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 549, in run_command
    self.handle_extension_command_error(ece)
  File "C:\Users\glenn\zephyrproject\.venv\Lib\site-packages\west\app\main.py", line 412, in handle_extension_command_error
    self.cmd.err(msg, fatal=True)
    ^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'err'
```

**Pull request merged**